docs/development/CUSTOM_RULES.md

```diff
@@ -0,0 +1,206 @@
+:::
+
+As long as you are using `@typescript-eslint/parser` as the `parser` in your ESLint configuration, custom ESLint rules generally work the same way for JavaScript and TypeScript code.
+The main two changes to custom rules writing are:
+
+- [AST Extensions](#ast-extensions): targeting TypeScript-specific syntax in your rule selectors
+- [Typed Rules](#typed-rules): using the TypeScript type checker to inform rule logic
+
+## AST Extensions
+
+`@typescript-eslint/estree` creates AST nodes for TypeScript syntax with names that begin with `TS`, such as `TSInterfaceDeclaration` and `TSTypeAnnotation`.
+These nodes are treated just like any other AST node.
+You can query for them in your rule selectors.
+
+This rule written in JavaScript bans interfaces that start with a lower-case letter:
+
+```js
+export const rule = {
+  create(context) {
+    return {
+      TSInterfaceDeclaration(node) {
+        if (/\p{Lu}/.test(node.id.name[0])) {
+          context.report({
+            messageId: 'uppercase',
+            node: node.id,
+          });
+        }
+      },
+    };
+  },
+  meta: {
+    docs: {
+      category: 'Best Practices',
+      description: 'Interface names should start with an upper-case letter.',
+    },
+    messages: {
+      uppercase: 'Start this name with an upper-case letter.',
+    },
+    type: 'suggestion',
+    schema: [],
+  },
+};
+```
+
+### Node Types
+
+TypeScript types for nodes exist in a `TSESTree` namespace exported by `@typescript-eslint/experimental-utils`.
+The above rule body could be better written in TypeScript with a type annotation on the `node`:
+
+```ts
+import { TSESTree } from '@typescript-eslint/experimental-utils';
+import * as eslint from 'eslint';
+
+export const rule: eslint.Rule.RuleModule = {
+  create(context) {
+    return {
+      TSInterfaceDeclaration(node: TSESTree.TSInterfaceDeclaration) {
+        // ...
+      },
+    };
+  },
+  meta: {
+    // ...
+  },
+};
+```
+
+An `AST_NODE_TYPES` enum is exported as well to hold the values for AST node `type` properties.
+`TSESTree.Node` is available as union type that uses its `type` member as a discriminant.
+
+For example, checking `node.type` can narrow down the type of the `node`:
+
+```ts
+import {
+  AST_NODE_TYPES,
+  TSESTree,
+} from '@typescript-eslint/experimental-utils';
+
+export function describeNode(node: TSESTree.Node): string {
+  switch (node.type) {
+    case AST_NODE_TYPES.ArrayExpression:
+      return `Array containing ${node.elements.map(describeNode).join(', ')}`;
+
+    case AST_NODE_TYPES.Literal:
+      return `Literal value ${node.raw}`;
+
+    default:
+      return '🤷';
+  }
+}
+```
+
+## Type Checking
+
+:::tip
+Read TypeScript's [Compiler APIs > Using the Type Checker](https://github.com/microsoft/TypeScript/wiki/Using-the-Compiler-API#using-the-type-checker) section for how to use a program's type checker.
```

Comment:
I really wish that TS had some half-decent docs on their API here.

Reply:
**Yes.** I'm planning on tweeting them after this goes live to ask for something we can link!